### PR TITLE
Split llvm package into constituent components

### PR DESCRIPTION
**Pull request by saltedcoffii**
As requested by uberhacker, splitting llvm. Right now only the llvm-11 package works, everything else is coming when they compile, mostly, as compiling takes me forever.

### CODE REVIEW

File: packages/glibc.rb [hunk 1]
```diff
@@ -167,6 +167,9 @@ def self.install
       system "localedef --prefix=#{CREW_DEST_DIR} -i tr_TR -f UTF-8 tr_TR.UTF-8"
       system "localedef --prefix=#{CREW_DEST_DIR} -i zh_CN -f GB18030 zh_CN.GB18030"
     end
+
+    FileUtils.mkdir "#{CREW_DEST_PREFIX}/include/malloc"
+    FileUtils.symlink "../malloc.h", "#{CREW_DEST_PREFIX}/include/malloc/malloc.h"
```
Comment (uberhacker) [inline packages/glibc.rb:172]: This would require new binaries and bumping the version.

Comment (saltedcoffii) [reply packages/glibc.rb:172]: Okay we can remove it. It's not acutely necessary by any means.

**Comment by saltedcoffii** [comment]
I couldn't get flang, openmp and polly to compile because gentoo doesn't have ebuilds for those packages and I've been basing a lot of the work on those. If someone can figure it out, or give me hints, let me know, otherwise it's ready to merge

**Comment by saltedcoffii** [comment]
I figured it out it has to do with `llvm-lir`

**Comment by saltedcoffii** [comment]
I got clang-tools-extra working but i still can't get flang, openmp and polly working

File: packages/llvm.rb [hunk 2]
```diff
@@ -1,179 +1,92 @@
...
   when 'aarch64', 'armv7l'
     # LLVM_TARGETS_TO_BUILD = 'ARM;AArch64;AMDGPU'
     LLVM_TARGETS_TO_BUILD = 'all'.freeze
-    @ARCH_C_FLAGS = "-fPIC -march=armv7-a -mfloat-abi=hard -ccc-gcc-name #{CREW_BUILD}"
-    @ARCH_CXX_FLAGS = "-fPIC -march=armv7-a -mfloat-abi=hard -ccc-gcc-name #{CREW_BUILD}"
+    @ARCH_C_FLAGS = "-fPIC  -pipe -march=armv7-a -mfloat-abi=hard -ccc-gcc-name #{CREW_TGT}"
+    @ARCH_CXX_FLAGS = "-fPIC -pipe -march=armv7-a -mfloat-abi=hard -ccc-gcc-name #{CREW_TGT}"
     @ARCH_LDFLAGS = ''
-    @ARCH_LTO_LDFLAGS = "#{@ARCH_LDFLAGS} -flto=thin"
+    @ARCH_LTO_LDFLAGS = "#{@ARCH_LDFLAGS} -flto=auto"
```
Comment (satmandu) [inline packages/llvm.rb:32]: `-flto=auto` only works for gcc.  This might have to be built with clang, which will require `-flto=thin`

**Comment by satmandu** [comment]
1. You're going to need a `llvm_stage1.rb ` to build llvm with LTO. (and that requires `-flto=thin` in the second stage.)
2. The ruby rebuild shouldn't be in this package.
3. LLVM is at [12.0 RC4](https://github.com/llvm/llvm-project/releases/tag/llvmorg-12.0.0-rc4), so maybe this should be targeted towards LLVM 12.0.

File: packages/llvm.rb (hunk 2 above)
Comment (saltedcoffii) [reply packages/llvm.rb:32]: it seems to build fine with gcc, and gcc produces faster binaries than clang.

**Comment by saltedcoffii** [comment]
1. I don't think llvm_stage1 is necessary. Llvm itself builds fine with gcc, and gcc produces faster binaries than clang anyway.
2. I didn't know the ruby package was in here, that was an accident.
3. A beta build of a core package might be a nightmare. I think we should just stick with this for now, and build LLVM 12 when it's out and stabilized for a while.

**Comment by satmandu** [comment]
Please test to see if you can build llvm with `-flto=auto`. You're replacing a working package file, and many many hours have gone into getting that to build on all architectures. LLVM and GCC are some of the packages where you absolutely need to test before submitting a package file. (Not trying to be pedantic... Just pointing out it's really easy to get a non-working compile... And LLVM compiles take a LONG time.)

I wasn't needling you about ruby, just pointing it out. 😅

Re: llvm 12, we had llvm 10 last year, IIRC... If there's an error building, we should submit an error message upstream. I mean we should get llvm 12 when 12.0 comes out, not roll out the RC.  Llvm's point releases (and GCC's too, are somewhat itinerant.). For instance there was never a point release of GCC 10.2.1, but everyone is distributing GCC 10.2.1 builds. (As are we.)


And thanks for the work you're putting in to get this functioning properly. I know what a pain in the ass this package can be!

**Comment by uberhacker** [comment]
Were you still working on this @saltedcoffii?  There are conflicts to resolve.

**Comment by saltedcoffii** [comment]
No, but I can work on it if it's wanted @uberhacker

**Comment by satmandu** [comment]
I need to build a standalone lld package which can be used to build the rest of the llvm packages.

It needs this patch:
```
  def self.patch
    # Fix gold linker issues killing mesa builds as per
    # https://reviews.llvm.org/D100624 and
    # https://bugs.llvm.org/show_bug.cgi?id=49915 fixed in 13.0 but also
    # hopefully making it to 12.0.1
    # llvm needs to be built with a lld that itself has been built with the following fix:
    system "sed -i 's#write16(buf, s.sym->versionId);#write16(buf, s.sym->isLazy() ? VER_NDX_GLOBAL : s.sym->versionId);#g' \
      lld/ELF/SyntheticSections.cpp"
  end
```

**Comment by saltedcoffii** [comment]
@satmandu lld is a really fast linker but it results in slower code. Are you sure there's no way around this? Perhaps just using llvm 12?

**Comment by satmandu** [comment]
> @satmandu lld is a really fast linker but it results in slower code. Are you sure there's no way around this? Perhaps just using llvm 12?

The problem is that the `lld` in llvm 11 and 12 is broken, and you need a working `lld` to properly build `llvm`. 

Once you compile `llvm` with a working `lld` the llvm libraries work right when being linked from gold linker.

I managed to get llvm 12 working right by compiling it twice for x86_64. 

This is what I'm running twice on other architectures to try to get a working lld build, since the issue is fixed in llvm 13. The plan is to try to get llvm 12 to build for i686 and armv7l with that.

(The llvm 12 I just built with this method for x86_64 was finally able to build mesa, so the process does appear to work eventually.)

```
require 'package'

class Lld < Package
  description 'The LLVM Project is a collection of modular and reusable compiler and toolchain technologies. The optional packages clang, lld, lldb, polly, compiler-rt, libcxx, libcxxabi, and openmp are included.'
  homepage 'http://llvm.org/'
  @_ver = '13.0.0-8a4'
  version @_ver
  license 'Apache-2.0-with-LLVM-exceptions, UoI-NCSA, BSD, public-domain, rc, Apache-2.0 and MIT'
  compatibility 'x86_64'
  source_url "https://github.com/llvm/llvm-project.git"
  git_hashtag '8a4ee3b39c149d6cf344f0804a572d5ff120bdff'
  #source_url "https://github.com/llvm/llvm-project/archive/llvmorg-#{@_ver}.tar.gz"
  #source_sha256 '8e6c99e482bb16a450165176c2d881804976a2d770e0445af4375e78a1fbf19c'

  depends_on 'ocaml' => :build
  depends_on 'pygments' => :build
  depends_on 'ccache' => :build
  depends_on 'elfutils' # R
  depends_on 'gcc11' # R

  case ARCH
  when 'aarch64', 'armv7l'
    # LLVM_TARGETS_TO_BUILD = 'ARM;AArch64;AMDGPU'
    # LLVM_TARGETS_TO_BUILD = 'all'.freeze
    @ARCH_C_FLAGS = "-ltinfow -fPIC -march=armv7-a -mfloat-abi=hard -ccc-gcc-name #{CREW_BUILD}"
    @ARCH_CXX_FLAGS = "-fPIC -march=armv7-a -mfloat-abi=hard -ccc-gcc-name #{CREW_BUILD}"
    @ARCH_LDFLAGS = ''
    @ARCH_LTO_LDFLAGS = "#{@ARCH_LDFLAGS} -flto=thin"
    # compiler-rt fails with
    # [2967/6683] Building CXX object projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonSymbolizer.armhf.dir/sanitizer_unwind_linux_libcdep.cpp.o
    # FAILED: projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonSymbolizer.armhf.dir/sanitizer_unwind_linux_libcdep.cpp.o
    # CCACHE_CPP2=yes CCACHE_HASHDIR=yes /usr/local/bin/ccache /usr/local/bin/clang++ -DHAVE_RPC_XDR_H=1 -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGEFILE_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Iprojects/compiler-rt/lib/sanitizer_common -I/usr/local/tmp/crew/llvm.20210420052043.dir/llvm-project-llvmorg-12.0.0/compiler-rt/lib/sanitizer_common -Iinclude -I/usr/local/tmp/crew/llvm.20210420052043.dir/llvm-project-llvmorg-12.0.0/llvm/include -I/usr/local/tmp/crew/llvm.20210420052043.dir/llvm-project-llvmorg-12.0.0/compiler-rt/lib/sanitizer_common/.. -fPIC -march=armv7-a -mfloat-abi=hard -ccc-gcc-name armv7l-cros-linux-gnueabihf -flto=thin -fuse-ld=lld -fvisibility-inlines-hidden -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wno-noexcept-type -Wnon-virtual-dtor -Wsuggest-override -Wno-comment -fdiagnostics-color -flto=thin -Wall -std=c++14 -O3 -DNDEBUG  -fPIC -fno-builtin -fno-exceptions -fomit-frame-pointer -funwind-tables -fno-stack-protector -fno-sanitize=safe-stack -fvisibility=hidden -fno-lto -O3 -gline-tables-only -nostdinc++ -fno-rtti -std=c++14 -MD -MT projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonSymbolizer.armhf.dir/sanitizer_unwind_linux_libcdep.cpp.o -MF projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonSymbolizer.armhf.dir/sanitizer_unwind_linux_libcdep.cpp.o.d -o projects/compiler-rt/lib/sanitizer_common/CMakeFiles/RTSanitizerCommonSymbolizer.armhf.dir/sanitizer_unwind_linux_libcdep.cpp.o -c /usr/local/tmp/crew/llvm.20210420052043.dir/llvm-project-llvmorg-12.0.0/compiler-rt/lib/sanitizer_common/sanitizer_unwind_linux_libcdep.cpp
    # clang-11: warning: argument unused during compilation: '-fuse-ld=lld' [-Wunused-command-line-argument]
    # /usr/local/tmp/crew/llvm.20210420052043.dir/llvm-project-llvmorg-12.0.0/compiler-rt/lib/sanitizer_common/sanitizer_unwind_linux_libcdep.cpp:63:3: error: use of undeclared identifier '_Unwind_VRS_Result'; did you mean '_Unwind_Resume'?
    # _Unwind_VRS_Result res = _Unwind_VRS_Get(ctx, _UVRSC_CORE,
    # ^~~~~~~~~~~~~~~~~~
    # _Unwind_Resume
    #LLVM_PROJECTS_TO_BUILD = 'clang;clang-tools-extra;libcxx;libcxxabi;libunwind;lldb;lld;polly;openmp'.freeze
  when 'i686'
    # LLVM_TARGETS_TO_BUILD = 'X86'.freeze
    # Because ld.lld: error: undefined symbol: __atomic_store
    # Polly demands fPIC
    @ARCH_C_FLAGS = '-ltinfow -latomic -fPIC'
    @ARCH_CXX_FLAGS = '-latomic -fPIC'
    # Because getting this error:
    # ld.lld: error: relocation R_386_PC32 cannot be used against symbol isl_map_fix_si; recompile with -fPIC
    # So as per https://github.com/openssl/openssl/issues/11305#issuecomment-602003528
    @ARCH_LDFLAGS = '-Wl,-znotext'
    @ARCH_LTO_LDFLAGS = "#{@ARCH_LDFLAGS} -flto=thin"
    #LLVM_PROJECTS_TO_BUILD = 'clang;clang-tools-extra;libcxx;libcxxabi;libunwind;lldb;compiler-rt;lld;polly;openmp'.freeze
  when 'x86_64'
    # LLVM_TARGETS_TO_BUILD = 'X86;AMDGPU'
    # LLVM_TARGETS_TO_BUILD = 'all'.freeze
    @ARCH_C_FLAGS = '-ltinfow -fPIC'
    @ARCH_CXX_FLAGS = '-fPIC'
    @ARCH_LDFLAGS = ''
    @ARCH_LTO_LDFLAGS = "#{@ARCH_LDFLAGS} -flto=thin"
    #LLVM_PROJECTS_TO_BUILD = 'clang;clang-tools-extra;libcxx;libcxxabi;libunwind;lldb;compiler-rt;lld;polly;openmp'.freeze
  end
  @ARCH_C_LTO_FLAGS = "#{@ARCH_C_FLAGS} -flto=thin -fuse-ld=lld"
  @ARCH_CXX_LTO_FLAGS = "#{@ARCH_CXX_FLAGS} -flto=thin -fuse-ld=lld"
  LLVM_PROJECTS_TO_BUILD = 'lld'.freeze

  # Using Targets 'all' for non-i686 because otherwise mesa complains.
  # This may be patched upstream as per
  # https://reviews.llvm.org/rG1de56d6d13c083c996dfd44a32041dacae037d66
  LLVM_TARGETS_TO_BUILD = 'all'.freeze

  def self.patch
    # This keeps system libraries from being linked in as dependencies on a build host image, such as in docker.
    # Uncomment the rest of this section if you need to.
    system "sudo rm /lib#{CREW_LIB_SUFFIX}/libncurses.so.5 || true"
    system "sudo rm /usr/lib#{CREW_LIB_SUFFIX}/libform.so || true"
    system "sudo ln -s #{CREW_LIB_PREFIX}/libncurses.so.6 /lib#{CREW_LIB_SUFFIX}/libncurses.so.5 || true"
    system "sudo ln -s #{CREW_LIB_PREFIX}/libform.so /usr/lib#{CREW_LIB_SUFFIX}/libform.so || true"
    # Fix gold linker issues killing mesa builds as per
    # https://reviews.llvm.org/D100624 and
    # https://bugs.llvm.org/show_bug.cgi?id=49915 fixed in 13.0 but also
    # hopefully making it to 12.0.1
    system "sed -i 's#write16(buf, s.sym->versionId);#write16(buf, s.sym->isLazy() ? VER_NDX_GLOBAL : s.sym->versionId);#g' \
      lld/ELF/SyntheticSections.cpp"
  end

  def self.build
    ############################################################
    puts "Building LLVM Targets: #{LLVM_TARGETS_TO_BUILD}".lightgreen
    puts "Building LLVM Projects: #{LLVM_PROJECTS_TO_BUILD}".lightgreen
    ############################################################
    ############################################################
    puts 'Setting compile to use python3'.lightgreen
    ############################################################
    system "grep -rl '#!.*python' | xargs sed -i '1s/python$/python3/'"

    ############################################################
    # puts "Downloading binutils #{BINUTILS_BRANCH} src to enable gold plugin build".lightgreen
    ############################################################
    # system "git config --global advice.detachedHead false"
    # As per https://github.com/SVF-tools/SVF/wiki/Install-LLVM-Gold-Plugin-on-Ubuntu
    # system "git clone --depth 1 --branch #{BINUTILS_BRANCH} git://sourceware.org/git/binutils-gdb.git binutils"

    Dir.mkdir 'builddir'
    Dir.chdir('builddir') do
      system "env PATH=#{CREW_LIB_PREFIX}/ccache/bin:#{CREW_PREFIX}/bin:/usr/bin:/bin LD=ld.lld \
        cmake -G Ninja \
        -DLLVM_ENABLE_LTO=Thin \
        -DCMAKE_C_COMPILER=$(which clang) \
        -DCMAKE_CXX_COMPILER=$(which clang++) \
        -DCMAKE_INSTALL_PREFIX=#{CREW_PREFIX} \
        -DLLVM_DEFAULT_TARGET_TRIPLE=#{CREW_BUILD} \
        -DCMAKE_C_COMPILER_TARGET=#{CREW_BUILD} \
        -DCMAKE_ASM_COMPILER_TARGET=#{CREW_BUILD} \
        -DLLVM_TARGETS_TO_BUILD=\'#{LLVM_TARGETS_TO_BUILD}' \
        -DCMAKE_BUILD_TYPE=Release \
        -DLLVM_LIBDIR_SUFFIX='#{CREW_LIB_SUFFIX}' \
        -DCMAKE_LINKER=$(which ld.lld) \
        -D_CMAKE_TOOLCHAIN_PREFIX=llvm- \
        -DLLVM_BUILD_LLVM_DYLIB=ON \
        -DLLVM_LINK_LLVM_DYLIB=ON \
        -DLLVM_CCACHE_BUILD=ON \
        -DLLVM_ENABLE_FFI=ON \
        -DLLVM_ENABLE_RTTI=ON \
        -DCMAKE_C_FLAGS='#{@ARCH_C_LTO_FLAGS}' \
        -DCMAKE_CXX_FLAGS='#{@ARCH_CXX_LTO_FLAGS}' \
        -DCMAKE_EXE_LINKER_FLAGS='#{@ARCH_LTO_LDFLAGS}' \
        -DLLVM_PARALLEL_LINK_JOBS=1 \
        -DPYTHON_EXECUTABLE=$(which python3) \
        -DLLVM_INSTALL_UTILS=ON \
        -DLLVM_BINUTILS_INCDIR='#{CREW_PREFIX}/include' \
        -DLLVM_OPTIMIZED_TABLEGEN=ON \
        -DLLVM_ENABLE_TERMINFO=ON \
        -DLLVM_ENABLE_PROJECTS='#{LLVM_PROJECTS_TO_BUILD}' \
        -DLIBUNWIND_C_FLAGS='-fno-exceptions -funwind-tables' \
        -DLIBUNWIND_CXX_FLAGS='-fno-exceptions -funwind-tables' \
        -DLIBUNWIND_SUPPORTS_FUNWIND_TABLES_FLAG=ON \
        -DLIBUNWIND_SUPPORTS_FNO_EXCEPTIONS_FLAG=ON \
        -DCOMPILER_RT_BUILD_SANITIZERS=OFF \
        -DCOMPILER_RT_BUILD_LIBFUZZER=OFF \
        -DCOMPILER_RT_BUILD_BUILTINS=ON \
        -DCOMPILER_RT_DEFAULT_TARGET_ONLY=ON \
        -Wno-dev \
        ../llvm"
      system 'ninja'
    end
  end

  def self.install
    Dir.chdir('builddir') do
      system "DESTDIR=#{CREW_DEST_DIR} ninja install"
      FileUtils.install 'clc', "#{CREW_DEST_PREFIX}/bin/clc", mode: 0o755
      FileUtils.install 'clc++', "#{CREW_DEST_PREFIX}/bin/clc++", mode: 0o755
      FileUtils.mkdir_p "#{CREW_DEST_LIB_PREFIX}/bfd-plugins"
      FileUtils.ln_s "lib#{CREW_LIB_SUFFIX}/LLVMgold.so", "#{CREW_DEST_LIB_PREFIX}/bfd-plugins/"
    end
  end

  def self.check
    Dir.chdir('builddir') do
      # system "ninja check-llvm || true"
      # system "ninja check-clang || true"
      # system "ninja check-lld || true"
    end
  end
end
```

**Comment by saltedcoffii** [comment]
Alright. I'll let you handle this pr from here on out since you seem to have more experience than I do. Just try to use GCC and ld.gold wherever possible

**Comment by satmandu** [comment]
I don't want this PR! I just want llvm working properly so I can get mesa compiled right. This whole situation is a PITA.

**Comment by saltedcoffii** [comment]
I'll leave this in the archives until llvm 13 comes out in that case.

**Comment by uberhacker** [comment]
This PR has a bunch of conflicts.  Can we continue this or has it been abandoned?

**Comment by uberhacker** [comment]
I don't blame you for giving up on this @saltedcoffii. LoL

**Pull request closed without merging**